### PR TITLE
Add a simple in-house XORWOW implementation

### PR DESCRIPTION
**Pull request by sethrj**
This adds an implementation of the `xorwow` pseudorandom number generator described in [Marsaglia 2003](https://www.jstatsoft.org/article/download/v008i14/916). The state advancement is identical between that paper, cuRAND, hipRAND, and this implementation; but the key difference between the them all is how they initialize the state from the seed. Marsaglia provides an arbitrary seed (123456789, etc.) that cuRAND and hipRAND both reproduce with the input 64-bit int seed = 0, but they both then (seemingly arbitrarily) hash that int with the various components of the xorwow state. As in Exnihilo/ExaSMR/Shift, Celeritas uses a host-side RNG to initialize 64-bit random seeds in the cuRAND/hipRAND case. In contrast, the new xorwow implementation simply fills the entire state with random bits.

To convert the sampled bits into single- and double-precision numbers, a new "32-bit canonical generator" helper class is added to generate uniform doubles on `[0, 1)` (which is the range provided by C++11 standard generators, unlike the `(0, 1]` generated by cuRAND/hipRAND). For doubles (which have 53-bit mantissas), this class simply combines two 32-bit samples into the low 53 bits of a 64-bit int, then multiplies by 2^{-53}.

With this pull request it's possible to run the CPU-only demo loop without CUDA or HIP installed. It paves the way for us to try out `stdpar` to implement the demo loop, since we no longer need a device-specific library. Additionally, since we better control the layout of the RNG state it also provides us opportunity to test whether coalescing the RNG state accesses improves performance.

### CODE REVIEW

**Comment by sethrj** [comment]
Disabling for the moment since I'm getting crashing/hangs in the CMS and testem3 problems...

File: src/random/XorwowRngEngine.hh [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+/*!
+ * Specialization of GenerateCanonical for XorwowRngEngine.
+ */
+template<class RealType>
+class GenerateCanonical<XorwowRngEngine, RealType>
+{
+  public:
+    //!@{
+    //! Type aliases
+    using real_type   = float;
```
Comment (amandalund) [inline src/random/XorwowRngEngine.hh:69]: Sneaky 😉
```suggestion
    using real_type   = RealType;
```

Comment (sethrj) [reply src/random/XorwowRngEngine.hh:69]: 🤦‍♂️thanks... Amanda to the rescue again!

**Comment by sethrj** [comment]
OK thanks to @amandalund for finding the problem! (At least the most immediate one: we still may need to do some stress testing on our RNG samples, because the crash seemed to happen because we sampled ~1.0 and had a past-the-end access error on an array.) The CMS loop works now without crashing.

The object file size will shrink because we're no longer forced to carry around 1MB of device memory for xorwow "discard" tables that we don't use, the state memory usage should be slightly smaller since we don't have to save space for the extra box-mueller normal samples, and our kernels no longer have "constant" memory because cuRAND forces us to keep some static __constant__ tables around for Poisson samples which we don't use.

**Comment by sethrj** [comment]
Not sure what's going on with the failures but it looks like a problem with the `fetnat05` machine... @dalg24 any ideas?
```
fetnat05 seems to be running inside container 81eb89003478331765299e4e9ddb8c3fc695fb5382e5b4a5e06cdf515500177b
$ docker run -t -d -u 0:0 -v persistent-tmp:/tmp/persistent -w /var/jenkins/workspace/Celeritas_PR-404 --volumes-from 81eb89003478331765299e4e9ddb8c3fc695fb5382e5b4a5e06cdf515500177b -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** celeritas/ci-focal-cuda11:2021-12-18 cat
ERROR: Timeout after 180 seconds
```

**Comment by dalg24** [comment]
Not sure either.  I just performed all the system update, reinstalled NVIDIA drivers and restarted.  Please go ahead and try again.

File: test/random/XorwowRngEngine.test.cc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    HostStore states(*params, num_seeds);
+    RngTally  tally;
+
+    for (unsigned int i = 0; i < num_seeds; ++i)
+    {
+        XorwowRngEngine rng(states.ref(), ThreadId{i});
+        for (unsigned int i = 0; i < num_samples; ++i)
+        {
+            tally(celeritas::generate_canonical(rng));
```
Comment (amandalund) [inline test/random/XorwowRngEngine.test.cc:207]: Also need a `tally.check()`, right?

**Review by amandalund — APPROVED** [review]
Looks great @sethrj!

Yeah, the first two failing assertions I saw were both because `generate_canonical` returned exactly 1:
```console
/home/alund/celeritas_project/celeritas/src/physics/grid/TwodSubgridCalculator.hh:96:
celeritas: internal assertion failed: y >= y_grid.front() && y < y_grid.back()
```
`generate_canonical()` == 1 --> sample from reciprocal distribution >= upper bound --> sampled energy is off grid

```console
/home/alund/celeritas_project/celeritas/app/demo-loop/generated/../LDemoLauncher.hh:169:
celeritas: internal assertion failed: step > 0
```
`generate_canonical()` == 1  --> sample from exponential distribution == 0 --> sampled mfp == 0

**Comment by sethrj** [comment]
Something's very wrong if the RNG is returning 1... I specifically intended for it to return numbers on `[0, 1)`, and even if the upper end was a closed range, I wouldn't expect any exact 1s to be encountered (since the odds should be `2^{53}:1` i.e. once per petascale run...) I'll dig into it.

**Comment by amandalund** [comment]
@sethrj it should be good now, that was before when double was being converted to float.

**Comment by sethrj** [comment]
Oh, ok: I thought you'd encountered new errors after that fix. I'm adding some more testing, anyway...

**Pull request merged**